File: data-structures/tree/dom-tree/tree.ts
interface Node {
  key: string
  children: Node[]
  addChild: (key: string) => Node
}
function createNode(key: string): Node {
  const children: Node[] = []

  const addChild = (key: string) => {
    const node = createNode(key)
    children.push(node)
    return node
  }

  return {
    key,
    children,
    addChild,
  }
}

type VisitFnType = (node: Node, level: number) => void

function createTree(rootKey: string) {
  const root = createNode(rootKey)

  const print = () => {
    let result = ""
    const traverse = (node: Node, fn: VisitFnType, level: number) => {
      fn(node, level)
      if (node.children.length > 0) {
        for (const n of node.children) {
          traverse(n, fn, level + 1)
        }
      }
    }
    // FN
    const addKeyToResult = (node: Node, level: number) => {
      result += result.length === 0 ? node.key : `\n${" ".repeat(level * 2)}${node.key}`
    }

    traverse(root, addKeyToResult, 0)
    return result
  }

  return {print, root}
}

const dom = createTree("html")
const head = dom.root.addChild("head")
const body = dom.root.addChild("body")
head.addChild("title")
const main = body.addChild("main")
main.addChild("h1")
main.addChild("section")
body.addChild("footer")
console.log(dom.print())

File: data-structures/tree/dom-tree/tree.go
package main

import (
	"fmt"
	"strings"
)

// AddChildFn func
// type AddChildFn func(string) Node

// Node struct
type Node struct {
	key      string
	children []*Node
}

// CreateNode func
func (n *Node) CreateNode(key string) *Node {
	return &Node{
		key:      key,
		children: []*Node{},
	}
}

// AddChild func
func (n *Node) AddChild(key string) *Node {
	node := n.CreateNode(key)
	n.children = append(n.children, node)
	return node
}

// Tree struct
type Tree struct {
	root *Node
}

// CreateTree func
func (t *Tree) CreateTree(key string) *Tree {
	var node Node
	t.root = node.CreateNode(key)
	return t
}

// AddKeyToResult func
type AddKeyToResult = func(result *string, node *Node, level int)

// GetRoot func
func (t *Tree) GetRoot() *Node {
	return t.root
}

func (t *Tree) print() string {
	res := ""
	traverse(&res, t.GetRoot(), addKeyToResult, 0)
	return res
}

func traverse(result *string, node *Node, fn AddKeyToResult, level int) {
	fn(result, node, level)
	if len(node.children) > 0 {
		for _, n := range node.children {
			traverse(result, n, fn, level+1)
		}
	}
}

func addKeyToResult(result *string, node *Node, level int) {
	nodeKey := fmt.Sprintf("%s", node.key)
	if len(*result) == 0 {
		*result += nodeKey
	} else {
		*result += "\n " + strings.Repeat(" ", level*2) + " " + node.key
	}
}

func main() {
	var dom Tree
	dom.CreateTree("html")
	body := dom.GetRoot().AddChild("body")
	footer := dom.GetRoot().AddChild("footer")
	footer.AddChild("ul --- social media list")
	main := body.AddChild("main")
	main.AddChild("h1 --- hello")
	fmt.Println(dom.print())
}
